### PR TITLE
Add Span overloads in TextRenderer

### LINKED ISSUE #3651: API Proposal: Add span overloads to `TextRenderer`
## Proposal

We should add span overloads to `TextRenderer` to allow for efficient text rendering. Fortunately the Win32 APIs in play here take a length so we can use span data directly- we've already refactored internally to use `ReadOnlySpan<char>`.

``` C#
namespace System.Windows.Forms
{
    public sealed class TextRenderer
    {
        // Existing:
        public static void DrawText(IDeviceContext dc, string text, Font font, Point pt, Color foreColor);
        public static void DrawText(IDeviceContext dc, string text, Font font, Point pt, Color foreColor, Color backColor);
        public static void DrawText(IDeviceContext dc, string text, Font font, Point pt, Color foreColor, Color backColor, TextFormatFlags flags);
        public static void DrawText(IDeviceContext dc, string text, Font font, Point pt, Color foreColor, TextFormatFlags flags);
        public static void DrawText(IDeviceContext dc, string text, Font font, Rectangle bounds, Color foreColor);
        public static void DrawText(IDeviceContext dc, string text, Font font, Rectangle bounds, Color foreColor, Color backColor);
        public static void DrawText(IDeviceContext dc, string text, Font font, Rectangle bounds, Color foreColor, Color backColor, TextFormatFlags flags);
        public static void DrawText(IDeviceContext dc, string text, Font font, Rectangle bounds, Color foreColor, TextFormatFlags flags);
        public static Size MeasureText(IDeviceContext dc, string text, Font font);
        public static Size MeasureText(IDeviceContext dc, string text, Font font, Size proposedSize);
        public static Size MeasureText(IDeviceContext dc, string text, Font font, Size proposedSize, TextFormatFlags flags);
        public static Size MeasureText(string text, Font font);
        public static Size MeasureText(string text, Font font, Size proposedSize);
        public static Size MeasureText(string text, Font font, Size proposedSize, TextFormatFlags flags);

        // 1-1 Proposal
        // Note that backColor and proposedSize can't be defaulted as they are Color.Empty and Size(int.MaxValue, intMaxValue).
        public static void DrawText(IDeviceContext dc, ReadOnlySpan<char> text, Font font, Point pt, Color foreColor);
        public static void DrawText(IDeviceContext dc, ReadOnlySpan<char> text, Font font, Point pt, Color foreColor, Color backColor);
        public static void DrawText(IDeviceContext dc, ReadOnlySpan<char> text, Font font, Point pt, Color foreColor, Color backColor, TextFormatFlags flags);
        public static void DrawText(IDeviceContext dc, ReadOnlySpan<char> text, Font font, Point pt, Color foreColor, TextFormatFlags flags);
        public static void DrawText(IDeviceContext dc, ReadOnlySpan<char> text, Font font, Rectangle bounds, Color foreColor);
        public static void DrawText(IDeviceContext dc, ReadOnlySpan<char> text, Font font, Rectangle bounds, Color foreColor, Color backColor);
        public static void DrawText(IDeviceContext dc, ReadOnlySpan<char> text, Font font, Rectangle bounds, Color foreColor, Color backColor, TextFormatFlags flags);
        public static void DrawText(IDeviceContext dc, ReadOnlySpan<char> text, Font font, Rectangle bounds, Color foreColor, TextFormatFlags flags);
        public static Size MeasureText(IDeviceContext dc, ReadOnlySpan<char> text, Font font);
        public static Size MeasureText(IDeviceContext dc, ReadOnlySpan<char> text, Font font, Size proposedSize);
        public static Size MeasureText(IDeviceContext dc, ReadOnlySpan<char> text, Font font, Size proposedSize, TextFormatFlags flags);
        public static Size MeasureText(ReadOnlySpan<char> text, Font font);
        public static Size MeasureText(ReadOnlySpan<char> text, Font font, Size proposedSize);
        public static Size MeasureText(ReadOnlySpan<char> text, Font font, Size proposedSize, TextFormatFlags flags);
    }
}
```

## Justification

Drawing and measuring text is a pretty common activity. This will allow significantly more efficient text rendering when new string allocations can be avoided (slicing other input, building off of a stack based character array, etc.). In conjunction with `IDeviceContext` on `PaintEventArgs` and other internal performance improvements, painting a string on a blank form can be done with < 15% of the allocations it used to take.

## Implementation Details

We'll match throwing behavior of the original APIs. The original APIs did not throw for null or empty strings, they just no-op.

### PR DESCRIPTION
**Pull request by JeremyKuhne**
This adds the span overloads per #3651. It also obsoletes `TextFormatFlags.ModifyString` and fully blocks it in the span overloads. We'll add a safe API for getting output strings in a future version.

Resolves #3651

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/3711)

### CODE REVIEW

File: src/System.Windows.Forms/src/Resources/SR.resx [hunk 1]
```diff
@@ -6717,6 +6776,9 @@ Stack trace where the illegal operation occurred was:
     <value>The ImageList control used by the ListView for ListViewGroup images in all views.</value>
   </data>
   <data name="ConstructorArgumentInvalidValueType" xml:space="preserve">
```
Comment (RussKie) [inline src/System.Windows.Forms/src/Resources/SR.resx:6778]: Please revert all changes above this line

**Comment by weltkante** [comment]
I think you mixed up the issue IDs, this does not resolve #3710 since no output-overloads are provided

**Comment by JeremyKuhne** [comment]
> I think you mixed up the issue IDs, this does not resolve #3710 since no output-overloads are provided

Hmm, I don't recall adding that. Interesting.

File: src/System.Windows.Forms/src/Resources/SR.resx (hunk 1 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/Resources/SR.resx:6778]: Why do we bother reverting this when the tool does it? Seems unnecessary.

File: src/System.Windows.Forms/src/Resources/SR.resx [hunk 2]
```diff
@@ -53,10 +53,10 @@
     <value>2.0</value>
   </resheader>
   <resheader name="reader">
-    <value>System.Resources.ResXResourceReader, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+    <value>System.Resources.ResXResourceReader, System.Windows.Forms, Version=42.42.42.42, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
   </resheader>
   <resheader name="writer">
-    <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+    <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=42.42.42.42, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
```
Comment (weltkante) [inline src/System.Windows.Forms/src/Resources/SR.resx:59]: will this be able to resolve when deployed with actually released version numbers?

Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/Resources/SR.resx:59]: I don't know, really. I'll take a look into it and roll it back for now.

**Comment by RussKie** [comment]
> > I think you mixed up the issue IDs, this does not resolve #3710 since no output-overloads are provided
> 
> Hmm, I don't recall adding that. Interesting.

Sorry, that'd be me... 🤦‍♂️ I thought it was related

**Comment by JeremyKuhne** [comment]
> Please annotate all new public API with xml-docs.

How precisely do you want to do that? These are identical to the existing APIs with an additional caveat that they block the setting. And do we need to block this for docs or can I follow up in a separate PR?

**Comment by RussKie** [comment]
> These are identical to the existing APIs with an additional caveat that they block the setting.

We still need to add /// comments, as these get imported into the docs repo (via a complicated pipeline). AFAIK if we don't have these from the get go we'll have to add them manually to xml files over in the docs repo, which is a much-much harder job to do.
@carlossanlop is this right?

**Comment by RussKie** [comment]
Docs: https://github.com/dotnet/docs/issues/20804

**Pull request merged**